src/resources/espnow.cc

```diff
@@ -0,0 +1,237 @@
+
+#include "../event_sources/system_esp32.h"
+
+#define ESPNOW_DATAGRAM_NUM 16
+#define ESPNOW_TX_WAIT_US   1000
+
+namespace toit {
+
+const int kInvalidESPNow = -1;
+
+// Only allow one instance of provisioning running.
+static  ResourcePool<int, kInvalidESPNow> espnow_pool(
+  0
+);
+
+static SemaphoreHandle_t tx_sem;
+static esp_now_send_status_t tx_status;
+static QueueHandle_t rx_queue;
+
+struct DataGram {
+  uint8_t mac[6];
+  int len;
+  uint8_t buffer[0];
+};
+
+class ESPNowResourceGroup : public ResourceGroup {
+ public:
+  TAG(ESPNowResourceGroup);
+
+  ESPNowResourceGroup(Process* process);
+  ~ESPNowResourceGroup();
+};
+
+ESPNowResourceGroup::ESPNowResourceGroup(Process* process)
+  : ResourceGroup(process) {
+  tx_sem = xSemaphoreCreateCounting(1, 0);
+  ASSERT(tx_sem);
+
+  rx_queue = xQueueCreate(ESPNOW_DATAGRAM_NUM, sizeof(void *));
+  ASSERT(rx_queue);
+}
+
+ESPNowResourceGroup::~ESPNowResourceGroup() {
+  tx_sem = NULL;
+}
+
+MODULE_IMPLEMENTATION(espnow, MODULE_ESPNOW)
+
+static void espnow_send_cb(const uint8_t *mac_addr, esp_now_send_status_t status)
+{
+  tx_status = status;
+  portBASE_TYPE ret = xSemaphoreGive(tx_sem);
+  if (ret != pdTRUE) {
+    // ESP_LOGE("ESPNow", "Failed to give TX semaphore");
+  }
+}
+
+static void espnow_recv_cb(const uint8_t *mac_addr, const uint8_t *data, int data_len)
+{
+  struct DataGram *datagram = (struct DataGram *)malloc(sizeof(struct DataGram) + data_len);
+  if (!datagram) {
+    // ESP_LOGE("ESPNow", "Failed to malloc datagram");
+  }
+
+  datagram->len = data_len;
+  memcpy(datagram->mac, mac_addr, 6);
+  memcpy(datagram->buffer, data, data_len);
+
+  portBASE_TYPE ret = xQueueSend(rx_queue, &datagram, 0);
+  if (ret != pdTRUE) {
+    free(datagram);
+    ESP_LOGE("ESPNow", "Failed to send RX queue");
+  }
+}
+
+PRIMITIVE(init) {
+  ARGS(int, mode, Blob, pmk);
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) {
+    ALLOCATION_FAILED;
+  }
+
+  int id = espnow_pool.any();
+  if (id == kInvalidESPNow) ALREADY_IN_USE;
+
+  ESPNowResourceGroup* group = _new ESPNowResourceGroup(process);
+  if (!group) {
+    espnow_pool.put(id);
+    MALLOC_FAILED;
+  }
+
+  FATAL_IF_NOT_ESP_OK(esp_netif_init());
+  wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
+  wifi_mode_t wifi_mode = mode ? WIFI_MODE_AP : WIFI_MODE_STA;
+
+  FATAL_IF_NOT_ESP_OK(esp_wifi_init(&cfg));
+  FATAL_IF_NOT_ESP_OK(esp_wifi_set_storage(WIFI_STORAGE_RAM));
+  FATAL_IF_NOT_ESP_OK(esp_wifi_set_mode(wifi_mode));
+  FATAL_IF_NOT_ESP_OK(esp_wifi_start());
+
+  FATAL_IF_NOT_ESP_OK(esp_now_init());
+  FATAL_IF_NOT_ESP_OK(esp_now_register_send_cb(espnow_send_cb));
+  FATAL_IF_NOT_ESP_OK(esp_now_register_recv_cb(espnow_recv_cb));
+  if (pmk.length() > 0) {
+    FATAL_IF_NOT_ESP_OK(esp_now_set_pmk(pmk.address()));
+  }
+
+  proxy->set_external_address(group);
+
+  return proxy;
+}
+
+PRIMITIVE(send) {
+  ARGS(Blob, addr, Blob, data, bool, wait);
+
+  xSemaphoreTake(tx_sem, 0);
```

Comment:
If you are taking a semaphore with a timeout of zero should you not check the return value?

Reply:
It takes semaphore here, mainly to `reset` the semaphore for next sending.